core/state_transition.go

```diff
@@ -232,45 +232,40 @@ func (self *StateTransition) TransitionDb() (ret []byte, requiredGas, usedGas *b
 	contractCreation := MessageCreatesContract(msg)
 	// Pay intrinsic gas
 	if err = self.useGas(IntrinsicGas(self.data, contractCreation, homestead)); err != nil {
-		return nil, nil, nil, InvalidTxError(err)
+		return nil, nil, nil, false, InvalidTxError(err)
 	}
 
 	vmenv := self.env
 	//var addr common.Address
+	var vmerr error
 	if contractCreation {
-		ret, _, err = vmenv.Create(sender, self.data, self.gas, self.gasPrice, self.value)
-		if homestead && err == vm.CodeStoreOutOfGasError {
+		ret, _, vmerr = vmenv.Create(sender, self.data, self.gas, self.gasPrice, self.value)
+		if homestead && vmerr == vm.CodeStoreOutOfGasError {
 			self.gas = big.NewInt(0)
 		}
 
-		if err != nil {
-			ret = nil
-			glog.V(logger.Core).Infoln("VM create err:", err)
+		if vmerr != nil {
+			glog.V(logger.Core).Infoln("VM create err:", vmerr)
 		}
 	} else {
 		// Increment the nonce for the next transaction
 		self.state.SetNonce(sender.Address(), self.state.GetNonce(sender.Address())+1)
-		ret, err = vmenv.Call(sender, self.to().Address(), self.data, self.gas, self.gasPrice, self.value)
-		if err != nil {
-			glog.V(logger.Core).Infoln("VM call err:", err)
+		ret, vmerr = vmenv.Call(sender, self.to().Address(), self.data, self.gas, self.gasPrice, self.value)
+		if vmerr != nil {
+			glog.V(logger.Core).Infoln("VM call err:", vmerr)
 		}
 	}
 
-	if err != nil && IsValueTransferErr(err) {
-		return nil, nil, nil, InvalidTxError(err)
-	}
-
-	// We aren't interested in errors here. Errors returned by the VM are non-consensus errors and therefor shouldn't bubble up
-	if err != nil {
-		err = nil
+	if vmerr != nil && IsValueTransferErr(vmerr) {
+		return nil, nil, nil, false, InvalidTxError(vmerr)
 	}
 
 	requiredGas = new(big.Int).Set(self.gasUsed())
 
 	self.refundGas()
 	self.state.AddBalance(self.env.Coinbase(), new(big.Int).Mul(self.gasUsed(), self.gasPrice))
 
-	return ret, requiredGas, self.gasUsed(), err
+	return ret, requiredGas, self.gasUsed(), vmerr != nil, err
```

Comment:
Seems kinda weird to return `requiredGas, when it's just `requiredGas = new(big.Int).Set(self.gasUsed())`...

Reply:
Cleaned up.